### ISSUE TITLE
sentry_environment is not backed up and breaks alerts logic

### ISSUE THREAD

**Issue by matthewbyrne** [getsentry/self-hosted#1965]
### Self-Hosted Version

22.12.0

### CPU Architecture

x86_64

### Docker Version

20.10.23

### Docker Compose Version

1.29.2

### Steps to Reproduce

```
mkdir /root/sentry
cd /opt/sentry; docker-compose run --rm -T -e SENTRY_LOG_LEVEL=CRITICAL web export > /root/sentry/backup.json
```

Parse the backup.json and you will not find any environments backed up, but rules etc reference the `envrironment_id`

### Expected Result

The sentry_environment is exported and restored via the backup and restore scripts

### Actual Result

`sentry_environment` is not backed up and thus not restored. This results in sentry alerts acting upon incorrect environment events, as the `sentry_environment` table is populated with environments in the order in which those environments are ingested into the database.

eg. An alert is configured against an environment called `staging` with `environment_id=3`, but when the backup is restored and no environments are set, the `sentry_environment` with id=3 could be set to any environment name depending on what the 3rd unique environment event was.

**Comment by hubertdeng123** [comment]
@matthewbyrne I noticed in the past you were encountering the `UniqueViolation ('dupllicate key value violates unique constraint...)` issue. Were you able to resolve that? This is something I'm not actively working on and as far as I know there were Deserialization errors in 22.7.0+ that you should've encountered during your backup/restore. I'd like to get more context on the issues you're seeing so I can help fix it

**Comment by hubertdeng123** [comment]
The Deserialization errors should be fixed now, just attempted this on the latest nightly images

**Comment by matthewbyrne** [comment]
Hi @hubertdeng123 .
To fix the UniqueViolation errors, we attempted a clean reinstall of Sentry self-hosted. This didn't fix the issue, but we noticed issues with our restore of the sentry data backup, which suggested events coming into Sentry before the system was restored were causing duplicate key violations. We truncated the affected table, and these errors ceased.

Today, we're doing yet another clean install of Sentry, but this time we're blocking traffic to `/api/*` (and allowing traffic to `/api/(0|1)` to allow us to configure integrations - Another table which is not backed up but is required by foreign keys upon restoration of a backup).

```
django.db.utils.IntegrityError: ForeignKeyViolation('insert or update on table "sentry_alertruletriggeraction" violates foreign key constraint "sentry_alertruletrig_integration_id_f43d7209_fk_sentry_in"\nDETAIL:  Key (integration_id)=(1) is not present in table "sentry_integration".\n')
```

Our backup file, which as you outline likely had deserialiazation errors, had to be updated. We wrote a script to do this, based on JSON objects which had a field `value`, and their `model` corresponding DB table had a `value` field of type `text`

```
 cat backup.json| jq '.[] | select(.fields.value !=null)' | grep model | sort -u                                                                                                         
  "model": "sentry.counter",
  "model": "sentry.incidentactivity",
  "model": "sentry.option",
  "model": "sentry.organizationoption",
  "model": "sentry.projectoption",
  "model": "sentry.useroption",
```

The script acted as follows (TLDR: value field values for the above models were wrapped in double quotes)
```
import json
FILENAME = "backup.json"

with open(FILENAME, "r") as f:
  data = json.load(f)

models_to_change = [
    "sentry.counter",
    "sentry.incidentactivity",
    "sentry.option",
    "sentry.organizationoption",
    "sentry.projectoption",
    "sentry.useroption"
]

for item in data:
  if item["model"] in models_to_change:
    if isinstance(item['fields']['value'], bool):
      item["fields"]["value"] = str(item['fields']['value']).lower()
    elif item['fields']['value'] is not None:
      item['fields']['value'] = str(item['fields']['value'])

with open('output.json', 'w') as f:
  json.dump(data, f)
```

**Comment by matthewbyrne** [comment]
FYI. We do see significant number of `duplicate key value violates unique constraint` in our postgres logs. After about 30 minutes of uptime:

```
      4 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_email_email_key"
    152 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_email_pkey"
      3 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_environmentreleas_organization_id_release__720a8cc8_uniq"
     37 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_eventuser_project_id_hash_377a63c0_uniq"
      1 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_featureadoption_organization_id_feature_id_8ca58f61_uniq"
      3 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_grouphash_project_id_hash_dee0ec7b_uniq"
      1 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_grouprelease_group_id_release_id_envi_044354c8_uniq"
      3 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_organizationonboa_organization_id_task_e08c2f3d_uniq"
      5 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_release_organization_id_version_65da5843_uniq"
      4 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_releaseprojectenv_project_id_release_id_en_1b470936_uniq"
      2 postgres_1                                  | ERROR:  duplicate key value violates unique constraint "sentry_release_project_project_id_release_id_44ff55de_uniq"
      1 postgres_1                                  | ERROR:  insert or update on table "sentry_alertruletriggeraction" violates foreign key constraint "sentry_alertruletrig_integration_id_f43d7209_fk_sentry_in"```

**Comment by hubertdeng123** [comment]
Thanks for the update @matthewbyrne . The deserialization should be resolved in the 23.2.0 release that just went out today. I've seen threads on the `duplicate key value violates unique constraint` such as [here](https://github.com/getsentry/sentry/issues/8004). Looks like it's a non issue during runtime and rather an improvement that can be made

**Comment by hubertdeng123** [comment]
I'm wondering what the best case to handle the UniqueViolation errors during the import process though 🤔

**Comment by hubertdeng123** [comment]
Thought about this, I think throwing a warning including instructions on how to handle the error instead of exiting on `duplicate key value violates unique constraint` during backup/restore is reasonable

**Comment by matthewbyrne** [comment]
Any thoughts on the `sentry_environment` table not being part of the backup?

**Comment by hubertdeng123** [comment]
I dropped a line internally to see if there's a reason why it's not a part of the backup. Seems safe to add, so I'll put up a PR for that :)

**Comment by hubertdeng123** [comment]
@matthewbyrne You mentioned performing a restore on a clean install of self-hosted. Just to confirm, did you attempt to restore data after events were already coming in? I don't think a clean install of self-hosted with a clean db should hit duplicate key violations

**Comment by hubertdeng123** [comment]
Created a new issue to investigate integrity errors during backup/restore

**Comment by hubertdeng123** [comment]
Just to make sure, you're hitting the foreign key violation after a successful restoration of data and trying to set up the integrations after?

**Comment by hubertdeng123** [comment]
PR for environment and environment project exports now merged! 
https://github.com/getsentry/sentry/pull/44852

**Comment by matthewbyrne** [comment]
> Just to make sure, you're hitting the foreign key violation after a successful restoration of data and trying to set up the integrations after?

Yes

**Comment by matthewbyrne** [comment]
> @matthewbyrne You mentioned performing a restore on a clean install of self-hosted. Just to confirm, did you attempt to restore data after events were already coming in? I don't think a clean install of self-hosted with a clean db should hit duplicate key violations

Correct. It was a clean install, but the endpoint for data ingestion was still open so events were ingested before the data restore was complete. This caused duplicate key violations on data import.

We resolved this by blocking traffic to `api/` (we allowed traffic to `api/0` and `api/1`) the restored the data before events made it into Sentry. This resulted in no duplicate key violations on data import - duplicate key violations still existed in postgres logs, but did not fail the import.
